// @gamecrash/frontend/src/components/BetPanel.tsx
import React from "react";

export function BetPanel(props: {
  preparedAmount: string;
  onAddChip: (value: string) => void;
  onClear: () => void;
  onPlaceBet: () => void;
  onCashout: () => void;
  canBet: boolean;
  canCashout: boolean;
  cashoutLabel?: string;
  loading: boolean;
  cashoutPulse?: boolean;
  placedAmount?: string | null;
  metaLabel?: string | undefined;
  bettingEndsIn?: number | null;
}) {
  const chips = ["100", "200", "500", "1000", "5000", "10000"];

  return (
    <div className="bet-dock">
      {props.bettingEndsIn !== null && props.bettingEndsIn !== undefined ? (
        <div className="dock-meta">
          <span className="ends">Betting ends in {props.bettingEndsIn}s</span>
        </div>
      ) : props.metaLabel ? (
        <div className="dock-meta">
          <span className="note">{props.metaLabel}</span>
        </div>
      ) : null}
      <div className="amount-row">
        <div className="amount-display" aria-label="prepared-amount">
          {Number(props.preparedAmount || "0").toLocaleString("en-US")}
        </div>
        <button
          className={`cta-primary ${props.canCashout ? "cashout" : props.canBet && Number(props.preparedAmount || "0") > 0 ? "ready" : "disabled"} ${props.cashoutPulse ? "cashout-pulse" : ""}`}
          data-testid="main-cta"
          onClick={props.canCashout ? props.onCashout : props.onPlaceBet}
          disabled={props.loading || (!props.canCashout && (!props.canBet || Number(props.preparedAmount || "0") <= 0))}
        >
          {props.canCashout ? props.cashoutLabel ?? "Cash Out Now" : "Place Bet"}
        </button>
      </div>

      <div className="chips-row">
        {chips.map((v) => (
          <button
            key={v}
            type="button"
            className="chip-btn"
            onClick={() => props.onAddChip(v)}
            disabled={props.loading || props.canCashout}
          >
            {v === "1000" ? "1K" : v === "5000" ? "5K" : v === "10000" ? "10K" : v}
          </button>
        ))}
        <button type="button" className="chip-btn clear" onClick={props.onClear} disabled={props.loading || props.canCashout}>
          CLR
        </button>
      </div>

      {props.placedAmount ? (
        <div className="bet-tray enter">
          <span className="icon" />
          <span className="txt">{Number(props.placedAmount).toLocaleString("en-US")}</span>
        </div>
      ) : null}
    </div>
  );
}
