// @gamecrash/frontend/src/components/CrashGame.tsx
import React, { useEffect, useMemo, useState } from "react";
import { runtimeConfig } from "../config/runtimeConfig";
import { apiClient } from "../lib/apiClient";
import { createSocketClient } from "../lib/socketClient";
import { useCrashAudio } from "../lib/useCrashAudio";
import { useRewardAudio } from "../lib/useRewardAudio";
import type { ActiveRound, BetRecord, CrashRealtimeEvent, HistoryRound, RoundStatus } from "../types/crash";
import { BalancePanel } from "./BalancePanel";
import { BetPanel } from "./BetPanel";
import { CrashScene } from "./CrashScene";
import { HistoryBar } from "./HistoryBar";
import { getMultiplierColor } from "../lib/multiplierColor";

export function CrashGame() {
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);
  const [connectionStatus, setConnectionStatus] = useState<"LIVE" | "OFFLINE">("OFFLINE");
  const [balanceMinor, setBalanceMinor] = useState("0");
  const [round, setRound] = useState<ActiveRound | null>(null);
  const [multiplier, setMultiplier] = useState("1.000000");
  const [betAmount, setBetAmount] = useState("1000");
  const [preparedAmount, setPreparedAmount] = useState("0");
  const [bet, setBet] = useState<BetRecord | null>(null);
  const [history, setHistory] = useState<HistoryRound[]>([]);
  const [countdownValue, setCountdownValue] = useState(5);
  const [historyExpanded, setHistoryExpanded] = useState(false);
  const [cashoutPulse, setCashoutPulse] = useState(false);
  const [settledCountdown, setSettledCountdown] = useState<number | null>(null);
  const [cooldownEndsAt, setCooldownEndsAt] = useState<number | null>(null);
  const [rewardEffect, setRewardEffect] = useState<{
    token: number;
    text: string;
    intensity: "small" | "medium" | "big";
    hue: string;
  } | null>(null);
  const [bettingCloseAtMs, setBettingCloseAtMs] = useState<number | null>(null);
  const [bettingEndsIn, setBettingEndsIn] = useState<number | null>(null);

  const roundStatus: RoundStatus | "IDLE" = round?.status ?? "IDLE";

  const canBet = useMemo(() => roundStatus === "BETTING_OPEN" && !!round?.roundId, [roundStatus, round]);
  const canCashout = useMemo(() => roundStatus === "IN_PROGRESS" && bet?.status === "ACTIVE", [roundStatus, bet]);
  const cashoutPreview = useMemo(() => {
    if (!canCashout || !bet?.amountMinor) return null;
    const amountMinor = BigInt(bet.amountMinor);
    const [wholePart, decimalPart = ""] = multiplier.split(".");
    const scaledMultiplier = BigInt(wholePart) * 1_000_000n + BigInt((decimalPart + "000000").slice(0, 6));
    const payout = (amountMinor * scaledMultiplier) / 1_000_000n;
    return payout.toString();
  }, [canCashout, bet, multiplier]);
  useCrashAudio(roundStatus, countdownValue);
  const rewardAudio = useRewardAudio();

  // Drive countdown from server-provided timestamp when available
  useEffect(() => {
    if (cooldownEndsAt) {
      const tick = () => {
        const remMs = Math.max(0, cooldownEndsAt - Date.now());
        setSettledCountdown(Math.ceil(remMs / 1000));
      };
      tick();
      const id = window.setInterval(tick, 250);
      return () => window.clearInterval(id);
    } else {
      setSettledCountdown(null);
    }
    return undefined;
  }, [cooldownEndsAt]);

  useEffect(() => {
    if (roundStatus === "BETTING_OPEN" || roundStatus === "IN_PROGRESS") {
      setCooldownEndsAt(null);
    }
  }, [roundStatus]);

  // Drive "betting ends in" countdown/progress from server timestamp
  useEffect(() => {
    if (!bettingCloseAtMs || roundStatus !== "BETTING_OPEN") {
      setBettingEndsIn(null);
      return;
    }
    const tick = () => setBettingEndsIn(Math.max(0, Math.ceil((bettingCloseAtMs - Date.now()) / 1000)));
    tick();
    const id = window.setInterval(tick, 250);
    return () => window.clearInterval(id);
  }, [bettingCloseAtMs, roundStatus]);

  useEffect(() => {
    if (!historyExpanded) return;
    const previous = document.body.style.overflow;
    document.body.style.overflow = "hidden";
    return () => {
      document.body.style.overflow = previous;
    };
  }, [historyExpanded]);

  async function refreshState() {
    const [state, wallet, historyResponse] = await Promise.all([
      apiClient.getState(),
      apiClient.getBalance(runtimeConfig.demoUserId),
      apiClient.getHistory()
    ]);

    setRound(state.activeRound);
    setBalanceMinor(wallet.balanceMinor);
    setHistory(historyResponse.rounds);

    if (!state.activeRound || state.activeRound.status === "BETTING_OPEN") {
      setMultiplier("1.000000");
    } else if ((state.activeRound.status === "CRASHED" || state.activeRound.status === "SETTLED") && state.activeRound.crashMultiplier) {
      setMultiplier(state.activeRound.crashMultiplier);
    }

    // Do NOT clear an ACTIVE bet during BETTING_OPEN; it should become active as the round starts.
    // Only clear when there is no active round and previous bet wasn't cashed out (end of cycle).
    if (!state.activeRound) {
      setBet((prev) => (prev?.status === "ACTIVE" ? { ...prev, status: prev.status } : prev));
    }
  }

  useEffect(() => {
    setLoading(true);
    refreshState()
      .catch((e) => setError(e instanceof Error ? e.message : "Failed to load state"))
      .finally(() => setLoading(false));

    const socket = createSocketClient();
    const handleReconnect = () => {
      setConnectionStatus("LIVE");
      refreshState().catch(() => undefined);
    };
    const handleReconnectError = () => {
      setConnectionStatus("OFFLINE");
      refreshState().catch(() => undefined);
    };

    const onRoundCreated = (evt: CrashRealtimeEvent) => {
      setRound((prev) => ({ roundId: evt.roundId ?? prev?.roundId ?? "", status: "BETTING_OPEN", crashMultiplier: null }));
      setMultiplier("1.000000");
      setBet(null);
    };

    const onBettingOpen = (evt: CrashRealtimeEvent) => {
      setRound((prev) => ({ roundId: evt.roundId ?? prev?.roundId ?? "", status: "BETTING_OPEN", crashMultiplier: null }));
      if (evt.bettingCloseAt) {
        const t = new Date(evt.bettingCloseAt).getTime();
        if (Number.isFinite(t)) setBettingCloseAtMs(t);
      }
    };

    const onRoundStarted = (evt: CrashRealtimeEvent) => {
      setRound((prev) => ({ roundId: evt.roundId ?? prev?.roundId ?? "", status: "IN_PROGRESS", crashMultiplier: null }));
    };

    const onTick = (evt: CrashRealtimeEvent) => {
      if (typeof evt.multiplier === "string") {
        setMultiplier(evt.multiplier);
      }
    };

    const onCashoutAccepted = (evt: CrashRealtimeEvent) => {
      setBet((prev) => {
        if (!prev) return prev;
        return {
          ...prev,
          status: "CASHED_OUT",
          payoutAmountMinor: evt.payoutAmountMinor ?? prev.payoutAmountMinor,
          cashoutMultiplier: evt.cashoutMultiplier ?? prev.cashoutMultiplier
        };
      });
      const payoutMinor = evt.payoutAmountMinor ?? "0";
      const payoutNumeric = Number(payoutMinor);
      const rewardText = Number.isFinite(payoutNumeric) ? payoutNumeric.toLocaleString("en-US") : payoutMinor;
      const m = Number(evt.cashoutMultiplier ?? "0");
      const intensity: "small" | "medium" | "big" = m > 10 ? "big" : m >= 3 ? "medium" : "small";
      const hue = `mc-${getMultiplierColor(m)}`;
      const token = Date.now();
      setRewardEffect({ token, text: rewardText, intensity, hue });
      rewardAudio.playReward(intensity);
      window.setTimeout(() => {
        setRewardEffect((prev) => (prev?.token === token ? null : prev));
      }, 1200);
      refreshState().catch(() => undefined);
    };

    const onRoundCrashed = (evt: CrashRealtimeEvent) => {
      setRound((prev) => ({
        roundId: evt.roundId ?? prev?.roundId ?? "",
        status: "CRASHED",
        crashMultiplier: evt.crashMultiplier ?? prev?.crashMultiplier ?? null
      }));
      if (typeof evt.crashMultiplier === "string") {
        setMultiplier(evt.crashMultiplier);
      }
      setBet((prev) => (prev?.status === "ACTIVE" ? { ...prev, status: "LOST" } : prev));
      if (evt.nextRoundStartsAt) {
        setCooldownEndsAt(new Date(evt.nextRoundStartsAt).getTime());
      }
    };

    const onRoundSettled = (evt: CrashRealtimeEvent) => {
      setRound((prev) => (prev ? { ...prev, status: "SETTLED" } : prev));
      if (evt.nextRoundStartsAt) {
        setCooldownEndsAt(new Date(evt.nextRoundStartsAt).getTime());
      }
      refreshState().catch(() => undefined);
    };

    socket.on("ROUND_CREATED", onRoundCreated);
    socket.on("BETTING_OPEN", onBettingOpen);
    socket.on("ROUND_STARTED", onRoundStarted);
    socket.on("MULTIPLIER_TICK", onTick);
    socket.on("CASHOUT_ACCEPTED", onCashoutAccepted);
    socket.on("ROUND_CRASHED", onRoundCrashed);
    socket.on("ROUND_SETTLED", onRoundSettled);
    socket.on("connect", handleReconnect);
    socket.on("reconnect", handleReconnect);
    socket.on("reconnect_error", handleReconnectError);
    socket.on("connect_error", handleReconnectError);

    return () => {
      socket.off("ROUND_CREATED", onRoundCreated);
      socket.off("BETTING_OPEN", onBettingOpen);
      socket.off("ROUND_STARTED", onRoundStarted);
      socket.off("MULTIPLIER_TICK", onTick);
      socket.off("CASHOUT_ACCEPTED", onCashoutAccepted);
      socket.off("ROUND_CRASHED", onRoundCrashed);
      socket.off("ROUND_SETTLED", onRoundSettled);
      socket.off("connect", handleReconnect);
      socket.off("reconnect", handleReconnect);
      socket.off("reconnect_error", handleReconnectError);
      socket.off("connect_error", handleReconnectError);
      socket.close();
    };
  }, []);

  function addChip(value: string) {
    const a = BigInt(preparedAmount || "0");
    const b = BigInt(value);
    setPreparedAmount((a + b).toString());
  }

  function clearPrepared() {
    setPreparedAmount("0");
  }

  async function onPlaceBet() {
    if (!round?.roundId) return;
    const amountToBet = preparedAmount && preparedAmount !== "0" ? preparedAmount : betAmount;
    if (!canBet) {
      setError("Betting is closed for this round");
      return;
    }
    setLoading(true);
    setError(null);

    try {
      const response = await apiClient.placeBet({ userId: runtimeConfig.demoUserId, roundId: round.roundId, amountMinor: amountToBet });
      setBet(response.bet);
      rewardAudio.playBetPlaced();
      await refreshState();
      setPreparedAmount("0");
    } catch (e) {
      setError(e instanceof Error ? e.message : "Failed to place bet");
    } finally {
      setLoading(false);
    }
  }

  async function onCashout() {
    if (!round?.roundId) return;
    setLoading(true);
    setError(null);
    setCashoutPulse(true);
    rewardAudio.playClick();
    window.setTimeout(() => setCashoutPulse(false), 180);

    try {
      const response = await apiClient.cashout({ userId: runtimeConfig.demoUserId, roundId: round.roundId });
      setBet(response.bet);
      await refreshState();
    } catch (e) {
      setError(e instanceof Error ? e.message : "Failed to cash out");
    } finally {
      setLoading(false);
    }
  }

  function formatMultiplier(value: string): string {
    const n = Number(value);
    if (!Number.isFinite(n)) return `${value}x`;
    if (n >= 100) return `${Math.round(n)}x`;
    if (n >= 10) return `${n.toFixed(1)}x`;
    return `${n.toFixed(2)}x`;
  }

  function multiplierClass(value: string): string {
    const n = Number(value);
    const hue = getMultiplierColor(n);
    return `m-${hue}`;
  }

  return (
    <div className="app-shell">
      <header className="topbar">
        <div className="brand-title">CRASH</div>
        <div className="topbar-right">
          <div className={`status-badge ${connectionStatus === "LIVE" ? "status-live" : "status-offline"}`}>
            <span className="status-dot" />
            {connectionStatus}
          </div>
          <BalancePanel balanceMinor={balanceMinor} />
        </div>
      </header>

      <HistoryBar
        rounds={history}
        expanded={historyExpanded}
        onToggle={() => setHistoryExpanded((x) => !x)}
        onClose={() => setHistoryExpanded(false)}
      />

      <section className="hero hero-flex">
        <CrashScene
          roundStatus={roundStatus}
          multiplier={multiplier}
          crashMultiplier={round?.crashMultiplier ?? null}
          onCountdownChange={setCountdownValue}
          displayMultiplier={formatMultiplier(multiplier)}
          multiplierClass={multiplierClass(multiplier)}
          rewardEffect={rewardEffect}
        />
      </section>

      <section className="action-strip bet-dock-wrap">
        <div className="card action-card">
          {/* Minimal: remove extra HUD/panels; scene + bottom dock only */}

          <BetPanel
            preparedAmount={preparedAmount}
            onAddChip={(value) => { rewardAudio.playClick(); addChip(value); }}
            onClear={clearPrepared}
            onPlaceBet={onPlaceBet}
            onCashout={onCashout}
            cashoutLabel={cashoutPreview ? `Cash Out ${Number(cashoutPreview).toLocaleString("en-US")}` : undefined}
            canBet={canBet}
            canCashout={canCashout}
            loading={loading}
            cashoutPulse={cashoutPulse}
            placedAmount={bet?.amountMinor ?? null}
            metaLabel={
              !canCashout && settledCountdown !== null && roundStatus === "SETTLED"
                ? `Next round in ${settledCountdown}s`
                : undefined
            }
            bettingEndsIn={roundStatus === "BETTING_OPEN" && bettingEndsIn !== null ? bettingEndsIn : null}
          />
        </div>
      </section>
    </div>
  );
}
